# Pandas

In [1]:
import numpy as np
import pandas as pd

## Métodos de lectura de Datos

* Leer json (como un diccionario)

In [2]:
pd.read_json("BBDD/USD.json").head()

base       date  time_last_updated      rates
AED  USD 2020-05-05         1588637050   3.672028
ARS  USD 2020-05-05         1588637050  66.819777
AUD  USD 2020-05-05         1588637050   1.553857
BGN  USD 2020-05-05         1588637050   1.787290
BRL  USD 2020-05-05         1588637050   5.524262

* Leer tabla copiada al portapapeles de tu computadora (después de darle Ctrl + C, pues)

Copia la tabla [acá](https://es.tab-tv.com/?p=1387).

In [3]:
pd.read_clipboard()

Processor   series   Vertical Segment Processor.1 Number  Launch   Year  \
0         i9x  Desktop   i9-7940X   Q3'17   FCLGA2066    3.1     4.4  19.25   
1         i9x  Desktop   i9-7960X   Q3'17   FCLGA2066    2.6     4.4     22   
2         i9x  Desktop   i9-7920X   Q3'17   FCLGA2066    2.9     4.3   16.5   
3         i9x  Desktop  i9-7980XE   Q3'17   FCLGA2066    2.6     4.4  24.75   
4         i9x  Desktop   i9-7900X   Q2'17   FCLGA2066    3.3     4.5  13.75   
..        ...      ...        ...     ...         ...    ...     ...    ...   
607        i3   Mobile       None    None        None   None     NaN   None   
608        i3   Mobile       None    None        None   None     NaN   None   
609        i3   Mobile       None    None        None   None     NaN   None   
610        i3   Mobile       None    None        None   None     NaN   None   
611        i3   Mobile       None    None        None   None     NaN   None   

    Socket Frequency  ... Power Watts Intel  HD Graphics Process Techn.  \
0    14/28       yes  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   
1    16/32       yes  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   
2    12/24       yes  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   
3    18/36       yes  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   
4    10/20       yes  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   
..     ...       ...  ...   ...   ...   ...  ..      ...     ...    ...   
607   None      None  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   
608   None      None  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   
609   None      None  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   
610   None      None  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   
611   None      None  ...   NaN   NaN   NaN NaN      NaN     NaN    NaN   

    Memory  Speed  (Mhz)  
0      NaN    NaN    NaN  
1      NaN    NaN    NaN  
2      NaN    NaN    NaN  
3      NaN    NaN    NaN  
4      NaN    NaN    NaN  
..     ...    ...    ...  
607    NaN    NaN    NaN  
608    NaN    NaN    NaN  
609    NaN    NaN    NaN  
610    NaN    NaN    NaN  
611    NaN    NaN    NaN  

[612 rows x 29 columns]

* Lectura de worksheet de Excel por hoja

In [4]:
pd.read_excel("BBDD/Mapeo_Colaborativo_rescateMX.xlsx", # sheet names
              sheet_name="Derrumbes - Oficial (datos.gob.").head() #Albergues - Oficial # Derrumbes - Oficial (datos.gob.

entidad      Delegación                 colonia                    lugar  \
0    CdMx  Álvaro Obregón                Victoria             Deportivo G3   
1    CdMx  Álvaro Obregón                 Tolteca  Salón de usos múltiples   
2    CdMx    Azcapotzalco    Azcapotzalco, Centro               Delegación   
3    CdMx    Azcapotzalco  San Martin Xochinahuac        CAIS Azcapotzalco   
4    CdMx    Azcapotzalco  San Martin Xochinahuac                      NaN   

           Tipo entrecalles                calle numero_exterior  Necesidades  \
0  Delegacional         NaN    Av. Escuadrón 201             s/n          NaN   
1  Delegacional         NaN  Calle 10 y Canarios             NaN          NaN   
2           NaN         NaN     Castilla Oriente             NaN          NaN   
3  SEDESO GCDMX         NaN  Avenida del Rosario             s/n          NaN   
4  Delegacional         NaN  Avenida del Rosario             s/n          NaN   

                                    link_google_maps  \
0  https://www.google.com.mx/maps/place/Gimnasio+...   
1  http://maps.google.com/maps?q=loc:19.3894066,-...   
2  http://maps.google.com/maps?q=loc:19.4838002,-...   
3  http://maps.google.com/maps?q=loc:19.4994028,-...   
4  http://maps.google.com/maps?q=loc:19.4994028,-...   

                                  dirección_completa        lat        lon  \
0    Av. Escuadrón 201 s/n, Victoria, Álvaro Obregón  19.387702 -99.199703   
1      Calle 10 y Canarios , Tolteca, Álvaro Obregón  19.389407 -99.195609   
2  Castilla Oriente , Azcapotzalco, Centro, Azcap...  19.483800 -99.184590   
3  Avenida del Rosario s/n, San Martin Xochinahua...  19.499403 -99.195565   
4  Avenida del Rosario s/n, San Martin Xochinahua...  19.499403 -99.195565   

  última_fecha_actualización  
0           2017-09-24T22:30  
1           2017-09-24T22:30  
2           2017-09-24T22:30  
3           2017-09-24T22:30  
4           2017-09-24T22:30

* Conexión a una base de datos SQL con pandas 
### ESTE CODIGO ES DE MUESTRA, LES DARA ERROR A MENOS QUE TENGAN UNA BASE DE DATOS LLAMADA ligamx, con usuario ligamx_scrapper y que este en localhost

In [5]:
import mysql.connector # importar la base de datos
import getpass

password = getpass.getpass("dame tu contraseña o te pego con mi guitarra") #'1234qwer'

cnx = mysql.connector.connect(
        user='ligamx_scrapper', 
        password=password, # note apures... nunca uso esa contrasenia :v
        host='localhost',
        database='ligamx')

df = pd.read_sql_query("SELECT * FROM ligamx.match;", con=cnx)
df.head()

dame tu contraseña o te pego con mi guitarra ········


match_date tournament_name match_name             home_team  home_score  \
0  2019-07-19   apertura 2019  Jornada 1                Puebla           1   
1  2019-07-19   apertura 2019  Jornada 1                 Atlas           1   
2  2019-07-20   apertura 2019  Jornada 1               Pachuca           1   
3  2019-07-20   apertura 2019  Jornada 1  Atlético de San Luis           0   
4  2019-07-20   apertura 2019  Jornada 1               América           4   

      away_team  away_score  
0  Club Tijuana           3  
1     FC Juárez           0  
2          León           3  
3         Pumas           2  
4     Monterrey           2

## Filtros con Numpy y Pandas

In [6]:
nba = pd.read_csv('BBDD/nba.csv')

Por buena práctica pasamos todos los nombres de las columnas a minúsculas:

In [7]:
nba.columns = nba.columns.str.lower()

Filtramos valores nulos y los desplegamos:

In [8]:
nba[nba.name.isnull()]

name team  number position  age height  weight college  salary
457  NaN  NaN     NaN      NaN  NaN    NaN     NaN     NaN     NaN

Tenemos un último renglón vacío, no sé por qué. Vamos a deshacernos de él:

In [9]:
nba.drop(nba[nba.name.isnull()].index, inplace = True)
nba.tail()

name       team  number position   age height  weight   college  \
452    Trey Lyles  Utah Jazz    41.0       PF  20.0   6-10   234.0  Kentucky   
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0    Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0       NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0       NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0    Kansas   

        salary  
452  2239800.0  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0

Solucionado.

Crear variable de primer nombre.

In [10]:
nba['first_name'] = nba.name.str.split(' ').str.get(0)

Crear la variable apellido:

In [11]:
nba['last_name'] = nba.name.str.split(' ').str.get(1)

Quedarnos sólo con los John.

In [12]:
nba[nba['first_name'] == 'John']

name                team  number position   age height  weight  \
2    John Holland      Boston Celtics    30.0       SG  27.0    6-5   205.0   
126  John Jenkins        Phoenix Suns    23.0       SG  25.0    6-4   215.0   
216   John Henson     Milwaukee Bucks    31.0       PF  25.0   6-11   229.0   
382     John Wall  Washington Wizards     2.0       PG  25.0    6-4   195.0   

               college      salary first_name last_name  
2    Boston University         NaN       John   Holland  
126         Vanderbilt    981348.0       John   Jenkins  
216     North Carolina   2943221.0       John    Henson  
382           Kentucky  15851950.0       John      Wall

¿Qué nombres se repiten más de tres veces?

In [13]:
(nba['first_name'].value_counts())[nba['first_name'].value_counts() > 3]

Jordan     7
James      7
Chris      6
Kevin      5
Mike       5
Kyle       5
Andre      4
Brandon    4
Tyler      4
Jeff       4
John       4
Nikola     4
Anthony    4
Tony       4
Andrew     4
C.J.       4
Josh       4
Name: first_name, dtype: int64

## Ejercicios

1. ¿Cuántos nombres con jota hay? Hint: utilizar el método *str.startswith*.

2. Utilizar el método np.in1d con las columnas first_name y last_name. ¿Qué te devuelve? Utiliza lo aprendido para saber cuántos nombres en la lista son también apellidos.

## Métodos de numpy

La forma más eficiente de usar filtros es con máscaras tal cual se muestra a continuación:

In [14]:
mask = (nba['age'] > 30) & (nba['number'] == 23)

In [15]:
nba.iloc[np.where(mask)]

name                 team  number position   age height  weight  \
169  LeBron James  Cleveland Cavaliers    23.0       SF  31.0    6-8   250.0   
303  Kevin Martin    San Antonio Spurs    23.0       SG  33.0    6-7   199.0   

              college      salary first_name last_name  
169               NaN  22970500.0     LeBron     James  
303  Western Carolina    200600.0      Kevin    Martin

Sólo LeBron y Kevin Martin tienen más de 30 años y usan el número 23.

Si queremos crear arreglos en numpy tenemos el método array:

In [16]:
np.array([1,5,7,69,'CoRoNaVairus'])

array(['1', '5', '7', '69', 'CoRoNaVairus'], dtype='<U12')

Para hacer arreglos con ceros:

In [17]:
np.zeros([10,10])

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

Si queremos arreglos con 1's:

In [18]:
np.ones(69)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1.])

arange y linspace hacen en esencia lo mismo con una sutil diferencia:

In [19]:
np.arange(0,10,1)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [20]:
np.linspace(0,10,50)

array([ 0.        ,  0.20408163,  0.40816327,  0.6122449 ,  0.81632653,
        1.02040816,  1.2244898 ,  1.42857143,  1.63265306,  1.83673469,
        2.04081633,  2.24489796,  2.44897959,  2.65306122,  2.85714286,
        3.06122449,  3.26530612,  3.46938776,  3.67346939,  3.87755102,
        4.08163265,  4.28571429,  4.48979592,  4.69387755,  4.89795918,
        5.10204082,  5.30612245,  5.51020408,  5.71428571,  5.91836735,
        6.12244898,  6.32653061,  6.53061224,  6.73469388,  6.93877551,
        7.14285714,  7.34693878,  7.55102041,  7.75510204,  7.95918367,
        8.16326531,  8.36734694,  8.57142857,  8.7755102 ,  8.97959184,
        9.18367347,  9.3877551 ,  9.59183673,  9.79591837, 10.        ])

Para generar números aleatorios con distribución normal:

In [21]:
np.random.randn(10)

array([ 0.14040188, -1.47797703, -0.22675928, -0.4304116 ,  0.06178693,
       -0.68393709, -2.19414893,  0.05311644, -0.34980808, -0.53105816])

Para generar números aleatorios dentro de un intervalo específico:

In [22]:
np.random.randint(1,70, size = 5)

array([55, 48, 32, 61, 47])

Generemos una matriz aleatoria:

In [23]:
a = np.array([np.random.randint(1,70, size=20)])
a

array([[20, 60, 47, 21,  5, 44, 22, 35, 34,  1, 61,  5, 12, 24, 11, 26,
        24, 42, 65, 64]])

In [24]:
np.reshape(a, (5,4))

array([[20, 60, 47, 21],
       [ 5, 44, 22, 35],
       [34,  1, 61,  5],
       [12, 24, 11, 26],
       [24, 42, 65, 64]])

Para obtener el valor máximo de un *dataframe*:

In [25]:
nba.iloc[np.argmax(nba['age'].values)]

name                 Tim Duncan
team          San Antonio Spurs
number                       21
position                      C
age                          40
height                     6-11
weight                      250
college             Wake Forest
salary                 5.25e+06
first_name                  Tim
last_name                Duncan
Name: 298, dtype: object

Tim Duncan es el más vejestorio

In [26]:
nba.iloc[np.argmin(nba['age'].values)]

name          Devin Booker
team          Phoenix Suns
number                   1
position                SG
age                     19
height                 6-6
weight                 206
college           Kentucky
salary         2.12784e+06
first_name           Devin
last_name           Booker
Name: 122, dtype: object

Devin Booker es el más joven

Seleccionamos por índice:

In [27]:
np.array([12,3,4,5,53])[0:]

array([12,  3,  4,  5, 53])

In [28]:
np.array([12,3,4,5,53])[3:]

array([ 5, 53])

In [29]:
np.array([12,3,4,5,53])[ : :2]

array([12,  4, 53])

# Ejercicios Numpy

-------------------------
1. Generar un numpy array con los elementos donde a[i] = b[i]

In [30]:
a = np.array([1,2,3,2,3,4,3,4,5,6])
b = np.array([7,2,10,2,7,4,9,4,9,8])

---------------------------
2. Normaliza los siguientes valores para que sus valores queden entre 0 y 1

In [31]:
c = abs(np.random.randn(40) * (20))

----------------------------
3. Filtra en la primera columna del numpy array aquellos valores que cumplan iris < 5, en la segunda iris > 1.5.

In [36]:
iris_2d = np.genfromtxt("BBDD/iris.csv", delimiter=',', dtype='float', usecols=[1,2], skip_header=1)
iris_2d[0:2]

array([[5.1, 3.5],
       [4.9, 3. ]])

-------------------------
4. Remplaza los valores nulos del numpy array dado por cero. Hint: usa np.isnan

In [37]:
iris_2d = np.genfromtxt("BBDD/iris.csv", delimiter=',', dtype='float', usecols=[0,1,2,3], skip_header=1)
iris_2d[np.random.randint(150, size=20), np.random.randint(4, size=20)] = np.nan

In [38]:
iris_2d[np.isnan(iris_2d)]

array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan])

In [39]:
iris_2d[np.isnan(iris_2d)] = 0

In [40]:
iris_2d[np.isnan(iris_2d)]

array([], dtype=float64)

## Manejo de Tablas
Pasemos a un poco de Data Wrangling

In [43]:
data = pd.DataFrame(np.arange(6).reshape((2, 3)),index=pd.Index(['Ohio', 'Colorado'], name='state'),
                    columns=pd.Index(['one', 'two', 'three'], name='number'))
data

number    one  two  three
state                    
Ohio        0    1      2
Colorado    3    4      5

Aplicaremos diversos metodos de los dataframes para transformar la informacion

In [47]:
result = data.stack() # pasa de DataFrame a Serie el atributo de nivel corresponde a que columna toma para indexarlo primero o en que orden
result

state     number
Ohio      one       0
          two       1
          three     2
Colorado  one       3
          two       4
          three     5
dtype: int32

In [49]:
result.unstack() # pasa de Serie a DataFrame el atributo de nivel corresponde a que index toma para columnizarlo

number    one  two  three
state                    
Ohio        0    1      2
Colorado    3    4      5

## Joins

### Tipos de Joins

![Alt text](Imagenes/joins.PNG)

In [58]:
df1 = pd.DataFrame({'key': ['b', 'b', 'a', 'c', 'a', 'a', 'b'], 'data1': range(7)})
df2 = pd.DataFrame({'key': ['a', 'b', 'd'],'data2': range(3)})

df3 = pd.DataFrame({'lkey': ['b', 'b', 'a', 'c', 'a', 'a', 'b'],'data1': range(7)})
df4 = pd.DataFrame({'rkey': ['a', 'b', 'd'],'data2': range(3)})

In [59]:
df1

key  data1
0   b      0
1   b      1
2   a      2
3   c      3
4   a      4
5   a      5
6   b      6

In [60]:
df2

key  data2
0   a      0
1   b      1
2   d      2

# Se pueden unir tablas de diversas maneras como vimos en la imagen

El metodo __merge__ nos permite unir dos tablas, ya sea por index o por columna

In [61]:
pd.merge(df1, df2) # por default une por columna que coincidan

key  data1  data2
0   b      0      1
1   b      1      1
2   b      6      1
3   a      2      0
4   a      4      0
5   a      5      0

In [62]:
pd.merge(df1, df2, on='key') # puedes indicarle la columna con el metodo on

key  data1  data2
0   b      0      1
1   b      1      1
2   b      6      1
3   a      2      0
4   a      4      0
5   a      5      0

In [63]:
pd.merge(df3, df4, left_on='lkey', right_on='rkey') # puedes indicarle si quieres que tome una columna en especifico del lado izquierdo o derecho

lkey  data1 rkey  data2
0    b      0    b      1
1    b      1    b      1
2    b      6    b      1
3    a      2    a      0
4    a      4    a      0
5    a      5    a      0

In [64]:
pd.merge(df1, df2, how='outer') # puedes cambiar el tipo de join por default es un inner join, aqui vemos un outer join

key  data1  data2
0   b    0.0    1.0
1   b    1.0    1.0
2   b    6.0    1.0
3   a    2.0    0.0
4   a    4.0    0.0
5   a    5.0    0.0
6   c    3.0    NaN
7   d    NaN    2.0

In [65]:
df1 = pd.DataFrame({'key': ['b', 'b', 'a', 'c', 'a', 'b'],'data1': range(6)})
df2 = pd.DataFrame({'key': ['a', 'b', 'a', 'b', 'd'],'data2': range(5)})

In [70]:
df1

key  data1
0   b      0
1   b      1
2   a      2
3   c      3
4   a      4
5   b      5

In [71]:
df2

key  data2
0   a      0
1   b      1
2   a      2
3   b      3
4   d      4

In [68]:
pd.merge(df1, df2, on='key', how='left') # left join

key  data1  data2
0    b      0    1.0
1    b      0    3.0
2    b      1    1.0
3    b      1    3.0
4    a      2    0.0
5    a      2    2.0
6    c      3    NaN
7    a      4    0.0
8    a      4    2.0
9    b      5    1.0
10   b      5    3.0

In [72]:
pd.merge(df1, df2, how='inner')

key  data1  data2
0   b      0      1
1   b      0      3
2   b      1      1
3   b      1      3
4   b      5      1
5   b      5      3
6   a      2      0
7   a      2      2
8   a      4      0
9   a      4      2

In [73]:
left = pd.DataFrame({'key1': ['foo', 'foo', 'bar'],'key2': ['one', 'two', 'one'],'lval': [1, 2, 3]})
right = pd.DataFrame({'key1': ['foo', 'foo', 'bar', 'bar'],'key2': ['one', 'one', 'one', 'two'],'rval': [4, 5, 6, 7]})


In [74]:
left

key1 key2  lval
0  foo  one     1
1  foo  two     2
2  bar  one     3

In [75]:
right

key1 key2  rval
0  foo  one     4
1  foo  one     5
2  bar  one     6
3  bar  two     7

In [76]:
pd.merge(left, right, on=['key1', 'key2'], how='outer') # puedes hacer join por tantas columnas quieras, y las ira uniendo en orden

key1 key2  lval  rval
0  foo  one   1.0   4.0
1  foo  one   1.0   5.0
2  foo  two   2.0   NaN
3  bar  one   3.0   6.0
4  bar  two   NaN   7.0

In [77]:
pd.merge(left, right, on='key1') # si se llega a repetir el nombre de uns columna le pone un sufijo... en este caso para la tabla de la derecha_x y para la izq _y

key1 key2_x  lval key2_y  rval
0  foo    one     1    one     4
1  foo    one     1    one     5
2  foo    two     2    one     4
3  foo    two     2    one     5
4  bar    one     3    one     6
5  bar    one     3    two     7

In [78]:
pd.merge(left, right, on='key1', suffixes=('_left', '_right')) # puedes cambiar el nombre de los sufijos

key1 key2_left  lval key2_right  rval
0  foo       one     1        one     4
1  foo       one     1        one     5
2  foo       two     2        one     4
3  foo       two     2        one     5
4  bar       one     3        one     6
5  bar       one     3        two     7

Atributos del comando merge

<img src='Imagenes/merge_atributes.PNG'>

## Ejercicio:
Ahora haz el merge por index

## Concatenado
Pegar tablas ya sea a nivel registro o columna

In [79]:
arr = np.arange(12).reshape((3, 4))
np.concatenate([arr, arr], axis=1)

array([[ 0,  1,  2,  3,  0,  1,  2,  3],
       [ 4,  5,  6,  7,  4,  5,  6,  7],
       [ 8,  9, 10, 11,  8,  9, 10, 11]])

In [80]:
s1 = pd.Series([0, 1], index=['a', 'b'])
s2 = pd.Series([2, 3, 4], index=['c', 'd', 'e'])
s3 = pd.Series([5, 6], index=['f', 'g'])

In [81]:
s1

a    0
b    1
dtype: int64

In [82]:
s2

c    2
d    3
e    4
dtype: int64

In [83]:
s3

f    5
g    6
dtype: int64

In [84]:
pd.concat([s1, s2, s3]) # concatenado a nivel registro, por default (axis=0)

a    0
b    1
c    2
d    3
e    4
f    5
g    6
dtype: int64

In [86]:
pd.concat([s1, s2, s3], axis=1, sort=False) # concatenado por columna (axis=1) # noten como rellena con nulos donde no hay info

0    1    2
a  0.0  NaN  NaN
b  1.0  NaN  NaN
c  NaN  2.0  NaN
d  NaN  3.0  NaN
e  NaN  4.0  NaN
f  NaN  NaN  5.0
g  NaN  NaN  6.0

In [88]:
s4 = pd.concat([s1 * 5, s3]) # una manera de modificar valores y concatenarlos
s4

a    0
b    5
f    5
g    6
dtype: int64

In [90]:
pd.concat([s1, s4], axis=1, sort=False) # concatenado a nivel columna

0  1
a  0.0  0
b  1.0  5
f  NaN  5
g  NaN  6

In [91]:
pd.concat([s1, s4], axis=1, join='inner') # puedes hacer concatenado con tipo de join, solo aquellos que coincidan en index

0  1
a  0  0
b  1  5

In [93]:
result = pd.concat([s1, s1, s3], keys=['one', 'two', 'three']) # concatenado en multiindex, creando un nivel 0 con tus llaves que corresponden a cada tabla
result

one    a    0
       b    1
two    a    0
       b    1
three  f    5
       g    6
dtype: int64

# Ejercicio
Replica los siguientes Joins con los dataframes siguientes

* Append por columnas

![Alt text](Imagenes/append_column.PNG)

* Append por filas

![Alt text](Imagenes/append_row.PNG)

* Merge por una llave

![Alt text](Imagenes/merge_1key.PNG)

* Merge por 2 llaves

![Alt text](Imagenes/merge_2keys.PNG)

* Merge inner join

![Alt text](Imagenes/merge_inner.PNG)

* Merge left join

![Alt text](Imagenes/merge_left.PNG)

* Merge outer join

![Alt text](Imagenes/merge_outer.PNG)

* Merge left join con sufijos

![Alt text](Imagenes/merge_left_suffix.PNG)

* Merge por index

![Alt text](Imagenes/merge_index.PNG)

In [94]:
raw_data = {
        'subject_id': ['1', '2', '3', '4', '5'],
        'first_name': ['Alex', 'Amy', 'Allen', 'Alice', 'Ayoung'], 
        'last_name': ['Anderson', 'Ackerman', 'Ali', 'Aoni', 'Atiches']}
df_a = pd.DataFrame(raw_data, columns = ['subject_id', 'first_name', 'last_name'])

raw_data = {
        'subject_id': ['4', '5', '6', '7', '8'],
        'first_name': ['Billy', 'Brian', 'Bran', 'Bryce', 'Betty'], 
        'last_name': ['Bonder', 'Black', 'Balwner', 'Brice', 'Btisan']}
df_b = pd.DataFrame(raw_data, columns = ['subject_id', 'first_name', 'last_name'])

raw_data = {
        'subject_id': ['1', '2', '3', '4', '5', '7', '8', '9', '10', '11'],
        'test_id': [51, 15, 15, 61, 16, 14, 15, 1, 61, 16]}
df_n = pd.DataFrame(raw_data, columns = ['subject_id','test_id'])

In [95]:
df_a

subject_id first_name last_name
0          1       Alex  Anderson
1          2        Amy  Ackerman
2          3      Allen       Ali
3          4      Alice      Aoni
4          5     Ayoung   Atiches

In [96]:
df_b

subject_id first_name last_name
0          4      Billy    Bonder
1          5      Brian     Black
2          6       Bran   Balwner
3          7      Bryce     Brice
4          8      Betty    Btisan

In [97]:
df_n

subject_id  test_id
0          1       51
1          2       15
2          3       15
3          4       61
4          5       16
5          7       14
6          8       15
7          9        1
8         10       61
9         11       16

-------
No abras el siguiente [link](https://chrisalbon.com/python/data_wrangling/pandas_join_merge_dataframe/), definitvamente no contiene las respuestas.

## Función de Reducción de Memoria

Un problema al que puede enfrentarse uno en la industria es que el *DataFrame* con el que se trabaje usé demasiada memoria y las operaciones que necesitemos efectuar sobre él sean en extremo lentas. Para resolver este problema recurriremos a una función de reducción de memoria implementada en un [Notebook](https://www.kaggle.com/gemartin/load-data-reduce-memory-usage) de Kaggle.

Empecemos por leer los ya conocidos datos de Iris:

In [98]:
df = pd.read_csv('BBDD/Iris.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB


Uno de las características que nos devuelve el método .info() es el uso de memoria que, según esto, es de 83.7 KB.
También notemos que por defecto las variables son leídas como enteros o flotantes de 64 bits (máxima resolución posible por una computadora de 64 bits (que es el caso de mi computadora)).

El método .memory_usage() regresa la memoria empleada en bytes por cada columna y el índice por default

In [99]:
df.memory_usage()

Index             128
Id               1200
SepalLengthCm    1200
SepalWidthCm     1200
PetalLengthCm    1200
PetalWidthCm     1200
Species          1200
dtype: int64

Si hacemos un .sum() sobre este objeto obtendremos todos los bytes usados por la memoria para almacenar el DataFrame:

In [100]:
print("Los bytes usados por la memoria para almacenar el DataFrame es:", str(df.memory_usage().sum()) + str(' bytes'))

Los bytes usados por la memoria para almacenar el DataFrame es: 7328 bytes


En la literatura se reporta que es más común trabajar en términos de MB, así que empecemos por convertir estos bytes a Megabytes. La conversión de bytes a MB es:

$1 MB = 1024^2 bytes$

Por lo que para saber los MB de memoria que usa nuestro Dataframe:

In [101]:
start_mem = df.memory_usage().sum()/1024**2
print('Dataframe usa', str(round(start_mem,2))+str(' MB de memoria'))

Dataframe usa 0.01 MB de memoria


Antes de tratar la función de reducción de memoria debemos revisar un detalle más: el método .iinfo() de numpy.

Este método recibe como parámetro los np.intX / np.floatX, es decir, los enteros/"flotantes" de X bits disponibles en numpy. Veamos esto más a fondo. 

Usemos el ejemplo de los enteros de 8 bits: 

In [102]:
np.int8

numpy.int8

Al pasar esto como argumento del método .iinfo() de numpy:

In [103]:
np.iinfo(np.int8)

iinfo(min=-128, max=127, dtype=int8)

np.iinfo nos da la información del objeto np.int8, a saber: son enteros de 8 bits con signo que van desde el entero -128 hasta el 127.

Si accedemos a los atributos max y min de este nuevo objeto obtenedremos los valores extremos de los enteros de 8 bits:

In [104]:
print("El valor mínimo posible de los enteros con signo de 8 bits es:", np.iinfo(np.int8).min, "y el valor máximo es:", np.iinfo(np.int8).max)

El valor mínimo posible de los enteros con signo de 8 bits es: -128 y el valor máximo es: 127


Para el caso de los flotantes, se debe usar el método np.fiinfo():

In [105]:
np.finfo(np.float16)

finfo(resolution=0.001, min=-6.55040e+04, max=6.55040e+04, dtype=float16)

**Tarea**: Investigar más a fondo eso de resolution y porqué pareciera no ser congruente con el mínimo y el máximo que nos está indicando.

In [106]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB


Entonces, ¿qué es lo que hace el código?

In [107]:
def reduce_mem_usage(df): # La función recibe un sólo parámetro que es el dataframe
    """ Itera sobre todas las columnas de un dataframe y modifica el tipo de dato 
        para recudicr el uso de memoria.
    """
    start_mem = df.memory_usage().sum() / 1024**2  # Se calcula la memoria empleada en MB
    print('Memoria en uso por el dataframe {:.2f} MB'.format(start_mem)) # Se imprime en pantalla la memoria empleada por el dataframe en MB
    
    for col in df.columns: # Iniciación de ciclo for sobre las columnas del dataframe
        col_type = df[col].dtype # Se guarda el tipo de la columna en la variable col_type
        
        if col_type != object: # Si la variable es numérica se entra a este if
            c_min = df[col].min() # Se calcula el valor mínimo de la columna
            c_max = df[col].max() # Se calcula el valor máximo de la columna
            if str(col_type)[:3] == 'int': # Se convierte la variable col_type a string y se toma los primeros tres caracteres, si son iguales a "int" entra al if
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max: # Si el valor mínimo de la columna es mayor al mínimo de los enteros con signo
                    df[col] = df[col].astype(np.int8) # de 8 bits y además el valor máximo de la columna es menor al máximo de enteros con signo de 8 bits, entonces...
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max: # En caso de no caer en el caso anterior, se repite el procedimiento para
                    df[col] = df[col].astype(np.int16) # determinar si es mejor que la variable sea entero de 16, 32 o de 64 bits 
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)  
            else: # En caso de que la columna numérica no sea de tipo entero, es un flotante y cae en este caso
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max: # Se sigue exactamente el mismo razonamiento que antes pero ahora con
                    df[col] = df[col].astype(np.float16) # flotantes
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)
        else: # En el caso de que la columna no sea numérica cae aquí
            df[col] = df[col].astype('category') # Todos los objetos son convertidos a categorías (¿Esto es recomendable en caso de tener que hacer parseo?)
            
    end_mem = df.memory_usage().sum() / 1024**2 # Una vez que se han modificado todas las columnas se vuelve a calcular la memoria utilizada para almacenar el dataframe
    print('Memoria en uso después de la optimización: {:.2f} MB'.format(end_mem)) # Se imprime la cantidad de memoria utilizada después de la optimización
    print('Decremento de {:.1f}%'.format(100 * (start_mem - end_mem) / start_mem)) # Porcentaje de reducción de memoria gracias a que se usó la función
    
    return df


def import_data(file):  # Función adicional
    """Crea dataframe y optimiza su memoria con la funcion reduce_mem_usage
    """
    df = pd.read_csv(file, parse_dates=True, keep_date_col=True) 
    df = reduce_mem_usage(df)
    return df

In [108]:
df = reduce_mem_usage(df)

Memoria en uso por el dataframe 0.01 MB
Memoria en uso después de la optimización: 0.00 MB
Decremento de 74.3%


In [109]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   Id             150 non-null    int16   
 1   SepalLengthCm  150 non-null    float16 
 2   SepalWidthCm   150 non-null    float16 
 3   PetalLengthCm  150 non-null    float16 
 4   PetalWidthCm   150 non-null    float16 
 5   Species        150 non-null    category
dtypes: category(1), float16(4), int16(1)
memory usage: 1.8 KB


En el caso de tener una variable con muchas categorías la función anterior puede no funcionar adecuadamente (e.g., trata de aplicar la función anterior a los datos del titanic). Para más información sobre uso de memoria por variables del tipo categoría presionar [aquí](https://pbpython.com/pandas_dtypes_cat.html).

## Comparación de Tiempos de Ejecución de Filtros con Numpy y con Pandas

In [89]:
# Tarda unos minutos...
DIM = 1 # Selecciona la dimension 1 o 2
N = 100 # Selecciona la cantidad de datos

if DIM == 1:
    npx = np.random.randn(N)
else:
    N = int(N**0.5)
    npx = np.random.randn(N, N)
dfx = pd.DataFrame(npx) # explicit loop (DIM == 1)
%timeit [x > 0 for x in npx]
# explicit loop (DIM == 2)                
%timeit [x > 0 for x in [y for y in npx]]   

%timeit npx > 0                             # numpy
%timeit np.where(npx > 0, True, False)      # np.where
%timeit dfx > 0                             # dataframe
%timeit dfx.applymap(lambda x: x > 0)       # applymap
%timeit dfx.apply(lambda x: x > 0, axis=0)  # apply, axis=0
%timeit dfx.apply(lambda x: x > 0, axis=1)  # apply, axis=1
%timeit dfx.pipe(lambda x: x > 0)           # pipe # INVESTIGAR!

86.2 µs ± 9.47 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
79.4 µs ± 3.86 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
2.46 µs ± 423 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
6.75 µs ± 587 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
280 µs ± 39.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
3.18 ms ± 450 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
2.67 ms ± 401 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
42.8 ms ± 3.57 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
266 µs ± 57.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


![Alt](Imagenes/numpy_efficient.PNG)

## Cuantiles y Momentos

### Cuantiles 

Valores de la variable que dividen a la distribución en un cierto número de partes de manera que cada una de ellas contiene el mismo número de frecuencia.

**Ejemplos** 

* Cuartiles ($Q_i$):  son valores de la variable que dividen a la distribución en 4 partes, cada una de las cuales engloba el 25% de las mismas. Se denotan de la siguiente forma: $Q_1$ es el primer cuartil que deja a su izquierda el 25% de los datos; $Q_2$ es el segundo cuartil que deja a su izquierda al 50% de los datos; $Q_3$ es el tercer cuartil que deja a su izquierda el 75% de los datos. ($Q_2$ = Mediana)

<img src = "Imagenes\cuantil.png">

* Deciles ($D_i$): son los valores de la variable que dividen a la distribución en diez partes iguales, cada una de las cuales engloba el 10% de los datos. En total se tienen 9 deciles ($Q_2 = D_5 = $ Mediana)

<img src="Imagenes\decil.png" width="500">

* Centiles o Percentiles ($P_i$): son los valores de la variable que dividen a la distribución en cien partes iguales, cada una de las cuales engloba el 1% de los datos. En total se tienen 99 percentiles ($Q_2 = D_5 = $ Mediana $ = P_{50}$)


## Momentos

Un "momento" es una medida cuantitativa específica de la forma de una función y es usada tanto en física como en estadística. Si la función es una distribución de probabilidad el momento cero es la probabilidad total (i.e. 1), el segundo momento central es la varianza, el tercer momento estandarizado es la asimetría y el cuarto momento estandarizado es la curtosis.

De forma general, el k-ésimo momento $\mu_k$ de una variable aleatoria X se define como:

$ \mu_k = E(X^k)$

El primer momento es la media aritmética y, en el caso de una variable aleatoria discreta se define como:

$ \mu_1 = E(X) = \frac{\Sigma^n x_i}{n}$

De forma general, el k-ésimo momento central $\mu_k$ de una variable aleatoria X se define como:

$ momento$ $central = E[(X-\mu_1)^k]$

La varianza es una cantidad específica que mide la dispersión de los valores de la variable respecto a la media aritmética. Cuanto mayor sea la varianza mayor dispersión existirá y por tanto la media aritmética tendrá menor representatividad.

Matemáticamente, la varianza $\sigma^2$se define como el segundo momento central. En términos de los momentos, puede demostrarse que:

$\sigma^2 = E[(X-\mu_1)^2] = E(X^2) -E(\mu_1^2) = \mu_2 - \mu_1^2$

Para el caso de una variable discreta se tiene que:

$\sigma^2 = \frac{1}{n} \Sigma_{i=1}^n (x_i^2 - \mu_1^2) $

Para el coeficiente de asimetría y la curtosis se define la variable estandarizada $X^*$ como:

$X^* = \frac{X-\mu_1}{\sigma}$

Así, el tercer momento central $\beta_3$ de la variable aleatoria estandarizada $X^*$ recibe el nombre de coefciente de asimetría:

$\beta_3 = E[(X^*)^3] = \frac{1}{\sigma^3}E[(X-\mu_1)^3]$

Para el caso de una variable discreta tenemos que:

$\beta_3 = \frac{1}{n \sigma^3} \Sigma_{i=1}^n(x_i-\mu_1)^3$ 


* Una distribución que es simétrica alrededor de su media aritmética tendrá un coeficiente de asimetría igual a cero.

* Una distribución que esté cargada a la derecha tendrá un coeficiente de asimetría positivo.

* Una distribución que esté cargada a la izquierda tendrá un coeficiente de asimetría negativo

<img src="Imagenes\skewness.jpg" width = 800>

El cuarto momento central $\beta_4$ de la variable aleatoria estandarizada $X^*$ recibe el nombre de curtosis:

$\beta_4 = E[(X^*)^4] = \frac{1}{\sigma^4}E[(X-\mu_1)^4]$

Bajo esta definición la curtosis de una distribución normal es igual a 3. Con esto en mente, se suele definir la curtosis bajo la misma definición de arriba, pero restándole un 3 para que la distribución normal tenga curtosis igual a cero:

$\beta_4 = \frac{1}{\sigma^4}E[(X-\mu_1)^4] -3$

Bajo esta definición se tiene que:

* Una distribución que normal tiene curtosis igual a cero y recibe el nombre de mesocúrtica.

* Una distribución más "puntiaguda" que la normal cuyos datos se aglomeran alrededor de la media tiene curtosis positiva y recibe el nombre de leptocúrtica.

* Una distribución "plana" tiene curtosis negativa y recibe el nombre de platicúrtica.

<img src="Imagenes\kurtosis.jpg">

Y para el caso de una variable discreta se tiene que:

$\beta_4 = \frac{1}{n \sigma^4} \Sigma_{i=1}^n(x_i-\mu_1)^4 - 3$ 

## Referencias

[1](https://mathcs.clarku.edu/~djoyce/ma217/moment.pdf) Moments and the moment generation function

[2](https://www.universoformulas.com/estadistica/descriptiva/asimetria-curtosis/) Asimetría y Curtosis

[3](http://www.economia.unam.mx/profesores/blopez/estadistica-descriptiva.pdf) Estadística y Conceptos. 

## Soluciones a ejercicios de numpy